services/deployment-agent/src/simcore_service_deployment_agent/git_url_watcher.py

```diff
@@ -74,48 +96,93 @@ class GitRepo:  # pylint: disable=too-many-instance-attributes, too-many-argumen
     password: str
     paths: List[Path]
     pull_only_files: bool
     directory: str = ""
 
+async def _checkout_repository(repo: GitRepo, tag: str = None):
+    if repo.pull_only_files:
+        await _git_checkout_files(repo.directory, repo.paths, tag)
+    else:
+        await _git_checkout_repo(repo.directory, tag)
 
-async def _init_repositories(repos: List[GitRepo]):
+async def _update_repository(repo: GitRepo):
+    if repo.pull_only_files:
+        await _git_pull_files(repo.directory, repo.paths)
+    else:
+        await _git_pull(repo.directory)
+
+async def _init_repositories(repos: List[GitRepo]) -> Dict:
+    description = {}
     for repo in repos:
         repo.directory = tempfile.TemporaryDirectory().name
         await _git_clone_repo(repo.repo_url, repo.directory, repo.branch, repo.username, repo.password)
-        if repo.pull_only_files:
-            await _git_checkout_files(repo.directory, repo.paths)
-        else:
-            await _git_checkout_repo(repo.directory)
-
-
-async def _check_repositories(repos: List[GitRepo]) -> Tuple[bool, str]:
-    change_detected = False
-    changes = ""
+        latest_tag = await _git_get_latest_matching_tag(repo.directory, repo.tags) if repo.tags else None
+        await _checkout_repository(repo, latest_tag)
+        sha = await _git_get_current_sha(repo.directory)
+        description[repo.repo_id] = f"{repo.repo_id}:{repo.branch}:{latest_tag}:{sha}" if latest_tag \
+            else f"{repo.repo_id}:{repo.branch}:{sha}"
+    return description
+
+async def _update_repo_using_tags(repo: GitRepo) -> Optional[str]:
+    log.debug("checking %s using tags", repo.repo_id)
+    # check if current tag is the latest and greatest
+    current_tag = await _git_get_current_matching_tag(repo.directory, repo.tags)
+    latest_tag = await _git_get_latest_matching_tag(repo.directory, repo.tags)
+    # there should always be a tag
+    assert latest_tag
+    if current_tag == latest_tag:
+        return
+    log.info("New tag detected: %s", latest_tag)
+
+    # get modifications
+    logged_changes = await _git_get_logs_tags(repo.directory, current_tag, latest_tag)
+    log.debug("%s tag changes: %s", latest_tag, logged_changes)
+
+    # checkout
+    await _checkout_repository(repo, latest_tag)
+    log.info("New tag %s  checked out", latest_tag)
+
+    # if the tag changed, an update is needed even if no files were changed
+    sha = await _git_get_current_sha(repo.directory)
+    return f"{repo.repo_id}:{repo.branch}:{latest_tag}:{sha}"
+
+async def _update_repo_using_branch_head(repo: GitRepo) -> Optional[str]:
+    modified_files = await _git_diff_filenames(repo.directory)
+    if not modified_files:
+        # no modifications
+        return
+    # get the logs
+    logged_changes = await _git_get_logs(repo.directory, repo.branch, repo.branch)
+    log.debug("Changelog:\n%s", logged_changes)
+    await _update_repository(repo)
+    # check if a watched file has changed

```

Comment:
I wonder if all the "human readable comments" shall be converted into ``log.debug``. It might help tracing the logic workflow